src/viam/robot/client.py

```diff
@@ -185,22 +187,37 @@ async def refresh(self):
         """
         response: ResourceNamesResponse = await self._client.ResourceNames(ResourceNamesRequest())
         resource_names: List[ResourceName] = list(response.resources)
-        if resource_names == self._resource_names:
-            return
-        manager = ResourceManager()
-        for rname in resource_names:
-            if rname.type not in [RESOURCE_TYPE_COMPONENT, RESOURCE_TYPE_SERVICE]:
-                continue
-            if rname.subtype == "remote":
-                continue
-            try:
-                manager.register(Registry.lookup_subtype(Subtype.from_resource_name(rname)).create_rpc_client(rname.name, self._channel))
-            except ResourceNotFoundError:
-                pass
         with self._lock:
+            if resource_names == self._resource_names:
+                return
+            for rname in resource_names:
+                if rname.type not in [RESOURCE_TYPE_COMPONENT, RESOURCE_TYPE_SERVICE]:
+                    continue
+                if rname.subtype == "remote":
+                    continue
+
+                if rname in self._manager.resources:
+                    res = self._manager.get_resource(ResourceBase, rname)
+                    if isinstance(res, ReconfigurableResourceRPCClientBase):
+                        res.reset_channel(self._channel)
```

Comment:
so if the resource list is different, do we always replace the channel even if we didn't lose connection?

Reply:
Updated so that we only replace the client/channel if the channel has changed